packages/runtime/container-runtime/src/compatUtils.ts

```diff
@@ -226,3 +266,85 @@ export function isValidMinVersionForCollab(
+ * For example, if a user sets the `enableGroupedBatching` option to true, but the minVersionForCollab
+ * is set to "1.0.0", then we should throw a UsageError since 1.x clients do not support batching.
+ * */
+export function validateRuntimeOptions(
+	minVersionForCollab: MinimumVersionForCollab,
+	runtimeOptions: Partial<ContainerRuntimeOptionsInternal>,
+): void {
+	getValidationForRuntimeOptions<RuntimeOptionsAffectingDocSchema>(
+		minVersionForCollab,
+		runtimeOptions as Partial<RuntimeOptionsAffectingDocSchema>,
+		runtimeOptionsAffectingDocSchemaConfigValidationMap,
+	);
+}
+
+/**
+ * Generic function to validate runtime options against the minVersionForCollab.
+ */
+export function getValidationForRuntimeOptions<T extends Record<string, unknown>>(
+	minVersionForCollab: SemanticVersion,
+	runtimeOptions: Partial<T>,
+	validationMap: ConfigValidationMap<T>,
+): void {
+	if (minVersionForCollab === defaultMinVersionForCollab) {
+		// If the minVersionForCollab is set to the default value, then we will not validate the runtime options
+		// This is to avoid disruption to users who have not yet set the minVersionForCollab value explicitly.
+		return;
+	}
+	// Iterate through each runtime option passed in by the user
+	for (const [passedRuntimeOption, passedRuntimeOptionValue] of Object.entries(
+		runtimeOptions,
+	) as [keyof T & string, T[keyof T & string]][]) {
+		// Skip if passedRuntimeOption is not in validation map
+		if (!(passedRuntimeOption in validationMap)) {
+			continue;
+		}
+
+		const requiredVersion = validationMap[passedRuntimeOption](passedRuntimeOptionValue);
+		if (requiredVersion !== undefined && gt(requiredVersion, minVersionForCollab)) {
+			throw new UsageError(
+				`Runtime option ${passedRuntimeOption}:${JSON.stringify(passedRuntimeOptionValue)} is not compatible with minVersionForCollab: ${minVersionForCollab}.`,
+			);
+		}
+	}
+}
+
+/**
+ * Helper function to map ContainerRuntimeOptionsInternal config values to
+ * minVersionForCollab in {@link runtimeOptionsAffectingDocSchemaConfigValidationMap}.
+ */
+export function configValueToMinVersionForCollab<
+	T,
+	Arr extends readonly [T, SemanticVersion][],
+>(configToMinVer: Arr): (value: T) => SemanticVersion | undefined {
+	const map = new Map(configToMinVer);
+	return (value: T) => {
+		// Try direct match first
+		if (map.has(value)) {
+			return map.get(value);
+		}
+		// If T is an object, check each key in the map
+		if (typeof value === "object") {
+			for (const [key, version] of map.entries()) {
+				if (
+					typeof key === "object" &&
+					key !== undefined &&
+					// If one of the properties in the object matches the passed in value,
+					// then we return the version. This is to support partial matches, for
+					// example { foo: "bar" } should match { foo: "bar", baz: "qux" }
+					Object.entries(key as Record<string, unknown>).some(
+						([k, v]) => (value as Record<string, unknown>)[k] === v,
+					)
+				) {
+					return version;
```

Comment:
The early return here means we don't easily support minimum versions for multiple independent options under an object key. This seems like something we will easily want in the future so I'd like the pattern for it established now. E.g. we add another option you can specify under `gcOptions`. Could you come up with something for that case and add test cases for it to your unit tests?